apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt

```diff
@@ -0,0 +1,345 @@
+	onComplete: (Result<Response, Exception>) -> Unit) {
+
+	val thing = graph[thingId]?.value
+
+	thing?.let {
+		val operation = thing.operations[operationId]
+
+		operation?.let {
+			operation.form?.let { form ->
+				if (form.properties.isEmpty()) {
+					requestProperties(form.id) {
+						form.properties = it
+						thing.operations[operationId] = operation
+
+						val attributes = fillFields(it)
+
+						performOperationRequest(operation.target, operation.method, attributes, onComplete)
+					}
+				}
+				else {
+					val attributes = fillFields(form.properties)
+
+					performOperationRequest(operation.target, operation.method, attributes, onComplete)
+				}
+			} ?: performOperationRequest(operation.target, operation.method, emptyMap(), onComplete)
+
+		} ?: onComplete(Result.of(null, { ApioException("Thing $it doesn't have the operation $operationId") }))
+
+	} ?: onComplete(Result.of(null, { ApioException("Thing not found") }))
+}
+
+fun performOperationRequest(url: String, method: String, attributes: Map<String, Any>,
+	onComplete: (Result<Response, Exception>) -> Unit) {
+	launch(UI) {
+		async(CommonPool) {
+
+			val json = Gson().toJson(attributes)
+			val request = createRequest(HttpUrl.parse(url), credential).newBuilder()
+				.method(method, RequestBody.create(MediaType.parse("application/json; charset=utf-8"), json))
+				.build()
+
+			val okHttpClient = OkHttpClient()
+
+			try {
+				Result.of(okHttpClient.newCall(request).execute())
+			} catch (e: Exception) {
+				Result.error(e)
+			}
+		}.await().let(onComplete)
+	}
+}
+
+fun requestProperties(url: String, onComplete: (List<Property>) -> Unit) {
+	val request = createRequest(HttpUrl.parse(url), credential)
+
+	launch(UI) {
+		async(CommonPool) {
+			val okHttpClient = OkHttpClient()
+			val response = okHttpClient.newCall(request).execute()
+
+			val json = response.body()?.string()
+
+			val mapType = TypeToken.getParameterized(Map::class.java, String::class.java, Any::class.java).type
+
+			val jsonObject = Gson().fromJson<Map<String, Any>>(json, mapType)
+
+			val supportedProperties = jsonObject["supportedProperty"] as List<Map<String, Any>>
+
+			supportedProperties.map {
+				val type = parseType(it["@type"])
+				val name = it["property"] as String
+				val required = it["required"] as Boolean
+				Property(type, name, required)
+			}
+		}.await().let(onComplete)
+	}
+}
+
+fun requestParseWaitLoop(url: HttpUrl,
+	fields: Map<String, List<String>>,
+	embedded: List<String>,
+	credentials: String?): Result<Thing, Exception> {
+	return try {
+		val response = request(url, fields, embedded, credentials)
+		parse(response)
+	} catch (e: IOException) {
+		Result.error(e)
+	}
+}
+
+private fun request(url: HttpUrl,
+	fields: Map<String, List<String>>,
+	embedded: List<String>, credentials: String?): Response {
+	val httpUrl = createUrl(url, fields, embedded)
+
+	if (credentials != null) {
+		credential = credentials
```

Comment:
the mixture of credential and credentials is strange

Reply:
Yes, it's strange, I renamed the property to `credentials` and also renamed the parameter to `newCredentials`.